lib/macs/cmac.dart

```diff
@@ -0,0 +1,287 @@
+ * </p><p>
+ * CMAC is a NIST recomendation - see 
+ * csrc.nist.gov/CryptoToolkit/modes/800-38_Series_Publications/SP800-38B.pdf
+ * </p><p>
+ * CMAC/OMAC1 is a blockcipher-based message authentication code designed and
+ * analyzed by Tetsu Iwata and Kaoru Kurosawa.
+ * </p><p>
+ * CMAC/OMAC1 is a simple variant of the CBC MAC (Cipher Block Chaining Message 
+ * Authentication Code). OMAC stands for One-Key CBC MAC.
+ * </p><p>
+ * It supports 128- or 64-bits block ciphers, with any key size, and returns
+ * a MAC with dimension less or equal to the block size of the underlying 
+ * cipher.
+ * </p>
+ */
+class CMac extends BaseMac {
+  static final FactoryConfig FACTORY_CONFIG = new DynamicFactoryConfig.suffix(
+      Mac,
+      "/CMAC",
+      (_, final Match match) => () {
+            BlockCipher cipher = new BlockCipher(match.group(1));
+            return new CMac(cipher, cipher.blockSize * 8);
+          });
+
+  Uint8List _poly;
+  Uint8List _ZEROES;
+
+  Uint8List _mac;
+
+  Uint8List _buf;
+  int _bufOff;
+  final BlockCipher _cipher;
+
+  final int _macSize;
+
+  Uint8List _Lu, _Lu2;
+
+  ParametersWithIV _params;
+
+  /**
+     * create a standard MAC based on a CBC block cipher (64 or 128 bit block).
+     * This will produce an authentication code the length of the block size
+     * of the cipher.
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     */
+//    CMac(BlockCipher cipher) : CMac(this._cipher, _cipher.blockSize * 8);
+
+  /**
+     * create a standard MAC based on a block cipher with the size of the
+     * MAC been given in bits.
+     * <p>
+     * Note: the size of the MAC must be at least 24 bits (FIPS Publication 81),
+     * or 16 bits if being used as a data authenticator (FIPS Publication 113),
+     * and in general should be less than the size of the block cipher as it reduces
+     * the chance of an exhaustive attack (see Handbook of Applied Cryptography).
+     *
+     * @param cipher        the cipher to be used as the basis of the MAC generation.
+     * @param macSizeInBits the size of the MAC in bits, must be a multiple of 8 and &lt;= 128.
+     */
+  CMac(BlockCipher cipher, int macSizeInBits)
+      : this._macSize = macSizeInBits ~/ 8,
+        this._cipher = new CBCBlockCipher(cipher) {
+    if ((macSizeInBits % 8) != 0) {
+      throw new ArgumentError("MAC size must be multiple of 8");
+    }
+
+    if (macSizeInBits > (_cipher.blockSize * 8)) {
+      throw new ArgumentError(
+          "MAC size must be less or equal to ${_cipher.blockSize * 8}");
+    }
+
+    _poly = lookupPoly(cipher.blockSize);
+
+    _mac = new Uint8List(cipher.blockSize);
+
+    _buf = new Uint8List(cipher.blockSize);
+
+    _ZEROES = new Uint8List(cipher.blockSize);
+
+    _bufOff = 0;
+  }
+
+  @override
+  String get algorithmName {
+    String blockCipherAlgorithmName = _cipher.algorithmName.split("/").first;
+    return "${blockCipherAlgorithmName}/CMAC";
```

Comment:
This is to get "AES/CMAC" instead of "AES/CBC/CMAC" (maybe it should be called that, though...)

Reply:
What is it called in Bouncy Castle?

Reply:
I think it would be called the same as the underlying cipher, "AES" (which seems incorrect):
```
public class CMac implements Mac                                  
{
    ...

    public String getAlgorithmName()
    {
        return cipher.getAlgorithmName();
    }
    ...
}
```
> `core/src/main/java/org/bouncycastle/crypto/macs/CMac.java`